python/ucxx/ucxx/_lib_async/continuous_ucx_progress.py

```diff
@@ -70,3 +77,133 @@ async def _progress_task(self):
             worker.progress()
             # Give other co-routines a chance to run.
             await asyncio.sleep(0)
+
+
+class BlockingMode(ProgressTask):
+    def __init__(
+        self,
+        worker: UCXWorker,
+        event_loop: asyncio.AbstractEventLoop,
+        progress_timeout: float = 1.0,
+    ):
+        """Progress the UCX worker in blocking mode.
+
+        The blocking progress mode ensure the worker is progress whenever the UCX
+        worker reports an event on its epoll file descriptor. In certain
+        circumstances the epoll file descriptor may not
+
+        Parameters
+        ----------
+        worker: UCXWorker
+            Worker object from the UCXX Cython API to progress.
+        event_loop: asyncio.AbstractEventLoop
+            Asynchronous event loop where to schedule async tasks.
+        progress_timeout: float
+            The timeout to sleep until calling checking again whether the worker should
+            be progressed.
+        """
+        super().__init__(worker, event_loop)
+
+        # Creating a job that is ready straight away but with low priority.
+        # Calling `await self.event_loop.sock_recv(self.rsock, 1)` will
+        # return when all non-IO tasks are finished.
+        # See <https://stackoverflow.com/a/48491563>.
+        self.rsock, wsock = socket.socketpair()
+        self.rsock.setblocking(0)
+        wsock.setblocking(0)
+        wsock.close()
+
+        epoll_fd = worker.epoll_file_descriptor
+
+        # Bind an asyncio reader to a UCX epoll file descriptor
+        event_loop.add_reader(epoll_fd, self._fd_reader_callback)
+
+        # Remove the reader and close socket on finalization
+        weakref.finalize(self, event_loop.remove_reader, epoll_fd)
+        weakref.finalize(self, self.rsock.close)
+
+        self.blocking_asyncio_task = None
+        self.last_progress_time = time.monotonic() - progress_timeout
+        self.asyncio_task = event_loop.create_task(self._timeout_progress(1.0))
+
+    def __del__(self):
+        """Cancel asynchronous blocking progress task.
+
+        Cancel asynchronouns blocking progress task.
+
+        .. warning::
+            This only works if the event loop is still running. If the event loop has
+            been closed before this runs the following error will be printed by the
+            interpreter on the standard output:
+
+            ```
+            Task was destroyed but it is pending!
+            ```
+        """
+        if self.event_loop is not None and self.event_loop.is_running():
+            if self.blocking_asyncio_task is not None:
+                self.call_soon_threadsafe(self.blocking_asyncio_task.cancel())
```

Comment:
**Problem**: Potential resource leak if `self.blocking_asyncio_task` is not properly cleaned up.

**Suggested Change**: Use `try`-`finally` block to ensure `self.blocking_asyncio_task` is cancelled even if an exception occurs.


**Severity (1 - 4)**: 2 - MINOR

**Line**: 145

```suggestion
                try:
                    self.call_soon_threadsafe(self.blocking_asyncio_task.cancel())
                finally:
                    self.blocking_asyncio_task = None
```

-----

<div align="right">
    <small><i>Generated by <b>Llama 3.1 405b</b></i></small>
</div>
<br />


<div align="right">
  <small>Was this helpful?</small>
  <a href="https://codecritic-feedback.nvidia.com?vote=1&comment_guid=5035951b-8953-4cc4-be77-49defd98b2b9&is_github=true" target="_blank">
    <button>👍</button>
  </a>
  <a href="https://codecritic-feedback.nvidia.com?vote=-1&comment_guid=5035951b-8953-4cc4-be77-49defd98b2b9&is_github=true" target="_blank">
    <button>👎</button>
  </a>
</div>